### ISSUE TITLE
Use SpecmaticJUnitSupport as a JUnit Extension

### ISSUE THREAD

**Issue by noelmccrory** [specmatic/specmatic#1073]
**Is your feature request related to a problem? Please describe.**
I am trying to implement contract tests in a existing, large Spring Boot service. We already have a large set of Component tests which all inherit from a base class which provides a lot of functionality, specifically around mocking external API calls. To use `SpecmaticJUnitSupport` our tests have to inherit from it which means that we can't use our existing base class.

**Describe the solution you'd like**
Would it be possible to provide the functionality in `SpecmaticJUnitSupport` as a JUnit extension?

**Describe alternatives you've considered**

1. Do not use `SpecmaticJUnitSupport` and instead invoke the Specmatic CLI to run the contract tests from a Junit test.
2. Refactor my existing tests so that use composition instead of inheritance (large effort)

**Additional context**
N/A

**Comment by yogeshnikam671** [comment]
Hi @noelmccrory, thanks for reaching out!

Junit5 extension is something that we also have been pondering over. 
We will get back to you on this.

**Comment by noelmccrory** [comment]
As a workaround, I replicated some of the functionality from the Specmatic CLI [TestCommand](https://github.com/znsio/specmatic/blob/392f6832e15d996ae53a39b4e00255dcaac8f6eb/application/src/main/kotlin/application/TestCommand.kt). I noticed that it uses `SpecmaticJUnitSupport` under the hood.

Here's what I ended up with (note, I'm using Spock):

```
    def "Verify API Spec adherence"() {
        given:
        System.setProperty("host", "localhost");
        System.setProperty("port", "8099");
        System.setProperty("bearerAuth", "abcdef");

        when:
        Launcher junitLauncher = LauncherFactory.create()

        var request = LauncherDiscoveryRequestBuilder.request()
                .selectors(DiscoverySelectors.selectClass(SpecmaticJUnitSupport.class))
                .build()
        junitLauncher.discover(request)

        var specmaticTestExecListener = new SpecmaticTestExecutionListener()
        junitLauncher.registerTestExecutionListeners(specmaticTestExecListener)
        junitLauncher.execute(request)

        then:
        specmaticTestExecListener.failureCount == 0
    }
```

Note: I had to duplicate [ContractExecutionListener](https://github.com/znsio/specmatic/blob/392f6832e15d996ae53a39b4e00255dcaac8f6eb/application/src/main/kotlin/application/test/ContractExecutionListener.kt) into my own repo (named `SpecmaticTestExecutionListener`), and remove `org.fusesource.jansi.AnsiConsole.systemInstall()` from `testPlanExecutionFinished()` as it wasn't playing nicely with Groovy, and not important for running the tests anyways.

**Comment by yogeshnikam671** [comment]
Hey @noelmccrory, Glad to see you come up with a workaround for this issue. Appreciate your initiative and thanks for sharing this.

However, it seems this will not give you great feedback in the IDE. 

So we will continue looking into this and come up with a solution such that it reduces your efforts and provides better feedback within IDE.

**Comment by yogeshnikam671** [comment]
@noelmccrory Also can you please share how you are using Specmatic in your projects? 
Would love to understand your use case in order to provide any suitable suggestions. 

You can setup a meeting with us through this [link](https://specmatic.in/contact-us/). Look forward to the conversation.
Thanks.

**Comment by yogeshnikam671** [comment]
Hi @noelmccrory, we spent some time in exploring if we can make SpecmaticJunitSupport a Junit Extension.
Turns out generating dynamic tests at runtime using Junit Extension is not possible. 
You can find the relevant details [here](https://github.com/junit-team/junit5/issues/371)

Although, we are trying out a different approach to solve your problem which is instantiating the dynamic tests via an interface instead of a class.
Here are the changes which are still [WIP](https://github.com/znsio/specmatic/compare/main...specmatic_junit_support_as_interface).

Let us know your thoughts on this approach.
We feel with this approach, your issue should be resolved. Although, please let us know if you have any feedback.

**Comment by noelmccrory** [comment]
Sounds good. How would that look when writing tests that use the new interface?

**Comment by yogeshnikam671** [comment]
Hey @noelmccrory, It will look exactly the same. 

The only difference will be that now instead of **extending the SpecmaticJunitSupport class**, you can **implement the SpecmaticContractTest interface**.
This will unblock you from extending other base classes.

Before : ```class TestClass extends SpecmaticJunitSupport { ... } ```

Now : ```class TestClass extends TestBaseClass implements SpecmaticContractTest { ... }  ```

Hope this clarifies. Let me know if you have any further questions.

**Comment by harikrishnan83** [comment]
@noelmccrory The [above solution](https://github.com/znsio/specmatic/issues/1073#issuecomment-2085003601) that @yogeshnikam671 mentioned is now available as part of release [1.3.14](https://github.com/znsio/specmatic/releases/tag/1.3.14). Please feel free to re-open the issue if you have any questions on this. Thanks.

**Comment by noelmccrory** [comment]
Thank you.

I forgot one very important detail - the current tests use Spock and the TestBaseClass extends Spock's Specification class. So I'm not able to make use of the new interface. 

I'll sure it'll come in useful in other repos which use JUnit. though.
